### ISSUE TITLE
`ddev pull` changes public/fileadmin to a directory (project type: typo3) 

### ISSUE THREAD

**Issue by jonaseberle** [drud/ddev#3216]
**Run a Diagnostic and Paste Link Here**
https://gist.github.com/jonaseberle/01aa87572b248d093109180e997d04b1

**Describe the bug**

In my case, `public/fileadmin` and `public/uploads` are symlinks.
After running `ddev pull`, the symlink `public/fileadmin` is replaced by an empty directory.

**To Reproduce**

```sh
ddev config --docroot=public --create-docroot --project-type=typo3

mkdir -p data/files/fileadmin data/files/uploads
ln --symbolic --relative data/files/* public/

ddev start

cat > .ddev/providers/test.yaml <<EOF
files_pull_command:
  command:  
EOF

ddev pull test --skip-db -y

# the former symlink `fileadmin` is now a folder
ls -als public/
```

**Expected behavior**

The symlink stays in place

**Version and configuration information (please complete the following information):**

- Host computer OS and Version: Linux (Ubuntu)
- ddev version information: v1.18.0-rc1

**Comment by rfay** [comment]
Thanks for the report. This is an unusual situation, right? What should the correct behavior be, just follow the symlink (if it's relative and also inside the filespace?) Pretty sure this behavior is a result of the lower-level utilities we use to copy directories, and I think they ignore symlinks.

Would it be possible to share your pull yaml?

**Comment by jonaseberle** [comment]
Hi Randy, the content of the pull yaml:

The used commands are those from ddev-contrib.
These commands don't do anything with `public/fileadmin`

```
environment_variables:
  target: live

auth_command:
  command: |
    ssh-add -l >/dev/null || ( echo "Please 'ddev auth ssh' before running this command." && exit 1 )

db_pull_command:
  command: |
    # Dumps and downloads from server
    /var/www/html/.ddev/commands/web/dump-db "${target}" -y
    # ddev only supports single-file-gzipped here currently so we gzip the downloaded SQL for ddev to consume it...
    cat /var/www/html/data/sql/struct/* /var/www/html/data/sql/records/* | gzip > /var/www/html/.ddev/.downloads/db.sql.gz

files_pull_command:
  command: |
    # This rsyncs to ./data/files/fileadmin and ./data/files/uploads because the default ddev location
    # is cleared before pulling
    # in our case, ./data/files/* is symlinked to appropriate locations in a post-start hook
    /var/www/html/.ddev/commands/web/dump-files "${target}" -y
```

What do I think it should do?
* It should not assume that fileadmin is the ONLY location where volatile filesystem data that needs downloading to dev instances is stored in a TYPO3 system
* I like to have a symlink there. One reason is that I have symlinks for volatile data (shared between releases) on production, too and I want to make sure that any code handles that well. - At least it should not change a symlink to something else.
* I do not put anything into the hardcoded expected path `.ddev/.downloads/files` - so it also should not do anything

Basically I think my problem is that `ddev pull` is extremely opinionated for TYPO3. 

I don't think it should need any knowledge about the used framework.

In my opinion a `pull` provider should be allowed to handle importing into database and putting files in the correct locations his/herself. This would open up so much more possibilities. Maybe an option like `automatic_import: false` would help.

**Comment by rfay** [comment]
> In my opinion a pull provider should be allowed to handle importing into database and putting files in the correct locations his/herself. This would open up so much more possibilities. Maybe an option like automatic_import: false would help.

I agree, in retrospect :) It would solve lots of problems.

Thanks for including your yaml, it will help in testing.

**Comment by rfay** [comment]
And `ddev pull` is opinionated to Drupal single-site... because I know it better than TYPO3. :)

I'm sure we can solve your symlink issue here, but it would be best to open a design suggestion issue for how you think pull can be altered, because your ideas have value.  And I know you're already thinking of how to do it in a nondisruptive way.

**Comment by rfay** [comment]
Maybe you were saying that `ddev pull` is too opinionated and limited for application to TYPO3. Could be. It also doesn't support private files on Drupal. Needs to grow up :)

**Comment by jonaseberle** [comment]
It's sufficient for the default case but currently it is limited (EDIT: also see last paragraph). So many other parts of ddev are so wonderfully designed to, while being extremely easy for the default case, being able to grow for nearly any need! That is so absolutely amazing!

As you mentioned there are sometimes custom, project-specific interesting locations for a web project.

Some examples I encountered in our projects with TYPO3:
* `public/uploads` - this has been an official location for (mostly) extension specific, volatile public files until TYPO3v9. 
* `var/cache/data/*` - a cache (default location for the TYPO3 Caching Framework API) that is expensive to rebuild, for example due to being a result of an external API (for example Geocoding results)
* `var/my_extension` - some extension's state data (for example private data)

In production, these will be "shared" folders, meaning that any release will use the same data by means of a symlink or mount to a common (inter-release) location. (Not all "shared" folders need downloading, though)

Adding some other ideas, I am trying to express it as a User Story:
As a developer, I would like to have `ddev pull`...
* allow **multiple, custom file system locations**
* allow **multiple databases**
* allow quick **"sync"ing of file system locations** (instead of using full .tgz) in order to save time/bandwidth
* stay **backwards compatible** with existing providers

... in order to pull external data into my local dev system in simple and complex projects.

Ideas for implementation:
* add these features to `ddev pull` by adding configuration, ... (complex)
* allow to disable current auto-handling of `/var/www/html/.ddev/.downloads/` per `*_pull_command`, for example by a config option `disable_auto_import: true` / `skip_import: true` (wording tbd)

Please advise: I can open a design suggestion issue.

Afterthought:
I've seen there is a `skipImportArg` already: https://github.com/drud/ddev/blob/master/pkg/ddevapp/provider.go#L123 - that's part of the [command line API](https://github.com/drud/ddev/blob/master/cmd/ddev/cmd/pull.go#L113) (yet not publicly documented I think) - that is something I can use as a workaround. 

It seems that the `skip_import` idea could be done by changing https://github.com/drud/ddev/blob/master/pkg/ddevapp/provider.go#L123 to 
```
if skipImportArg || p.FilesPullCommand.skip_import {
```
Taking that afterthought into account it is not "actually" limited ;) The limitation is only on the configuration side, the runtime side allows customization already.

**Comment by infabo** [comment]
Is this issue somehow resolved? Or is there a follow-up issue somwhere? I face this same issue for months now.

**Comment by rfay** [comment]
Yes, 
* https://github.com/drud/ddev/pull/3373

Looks like you got it.
